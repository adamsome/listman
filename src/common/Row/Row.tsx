/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import React from 'react'
import FlexBox from '../Flexbox'
import { useTheme } from '../theming'
import withButtonProps from '../with-button-props'

type Props = typeof defaultProps & {
  onClick?: (event: React.MouseEvent) => void
  children: React.ReactNode
  /**
   * Set to whether the row is first in a collection to manually control
   * borders. `last` must be set as well. If undefined, component will use
   * css selectors (e.g. `first-of-type`) to control borders.
   */
  first?: boolean
  /**
   * Set to whether the row is last in a collection to manually control
   * borders. `first` must be set as well. If undefined, component will use
   * css selectors (e.g. `first-of-type`) to control borders.
   */
  last?: boolean
}

const defaultProps = {}

const Row = (props: Props) => {
  const { first, last, ...rest } = withButtonProps(props)
  const { onClick } = rest
  const theme = useTheme()

  const manualBorders = first != null && last != null
  const manualFirst = manualBorders && first
  const manualLast = manualBorders && last

  return (
    <FlexBox
      css={css`
        background-color: ${theme.bg.content};
        border: 1px solid ${theme.border.content};
        border-top-left-radius: ${manualFirst && theme.space.borderRadius};
        border-top-right-radius: ${manualFirst && theme.space.borderRadius};
        border-bottom-left-radius: ${manualLast && theme.space.borderRadius};
        border-bottom-right-radius: ${manualLast && theme.space.borderRadius};
        border-bottom-width: ${manualLast ? '1px' : 0};
        &:first-of-type {
          border-top-left-radius: ${!manualBorders && theme.space.borderRadius};
          border-top-right-radius: ${!manualBorders &&
            theme.space.borderRadius};
        }
        &:last-of-type {
          border-bottom-left-radius: ${!manualBorders &&
            theme.space.borderRadius};
          border-bottom-right-radius: ${!manualBorders &&
            theme.space.borderRadius};
          border-bottom-width: ${!manualBorders && '1px'};
        }
        color: ${theme.body};
        cursor: ${onClick ? 'pointer' : 'auto'};
        font-size: ${theme.font.size.body};
        outline: none;
        overflow: hidden;
        text-align: start;

        &:focus {
          box-shadow: 0 0 0 2px ${theme.border.focus}, ${theme.shadow};
        }
      `}
      {...rest}
    />
  )
}

Row.defaultProps = defaultProps

export default Row
