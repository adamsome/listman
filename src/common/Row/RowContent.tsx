import styled from '@emotion/styled'
import React from 'react'

type Props = typeof defaultProps & {
  children: React.ReactNode
}

const defaultProps = {}

const RowContent = (props: Props) => {
  const { children, ...rest } = props

  return (
    <Wrapper {...rest}>
      <Content>{children}</Content>
    </Wrapper>
  )
}

const Wrapper = styled.div`
  flex: 1;
  margin-right: 1rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
  &:first-of-type {
    margin-left: 1rem;
  }
`

const Content = styled.div`
  display: block;
`

RowContent.defaultProps = defaultProps

export default RowContent
