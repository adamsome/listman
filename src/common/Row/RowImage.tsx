/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import React from 'react'
import styled from '../theming/styled'

type Props = typeof defaultProps &
  React.PropsWithoutRef<JSX.IntrinsicElements['img']> & {}

const defaultProps = {}

const RowImage = (props: Props) => {
  // TODO: Theme margins
  return (
    <ImageWrapper>
      <Image {...props} />
    </ImageWrapper>
  )
}

const ImageWrapper = styled.div`
  margin-left: ${props => props.theme.space.inset};
  margin-right: ${props => props.theme.space.inset};
  &:first-of-type {
    margin-left: 0;
  }
  &:last-of-type {
    margin-right: 0;
  }
`

const Image = (props: Props) => (
  <img
    css={css`
      display: block;
      max-width: 100%;
    `}
    alt={props.alt || ''}
    {...props}
  />
)

RowImage.defaultProps = defaultProps

export default RowImage
