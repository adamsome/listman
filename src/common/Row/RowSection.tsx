/** @jsx jsx */
import { css, jsx } from '@emotion/core'

type RowSectionProps = typeof defaultProps &
  React.PropsWithoutRef<JSX.IntrinsicElements['div']> & {}

const defaultProps = {}

const RowSection = (props: RowSectionProps) => (
  <div
    css={css`
      margin-right: 1rem;
      padding-top: 1rem;
      padding-bottom: 1rem;
    `}
    {...props}
  />
)

RowSection.defaultProps = defaultProps

export default RowSection
