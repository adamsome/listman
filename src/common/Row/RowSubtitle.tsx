import React from 'react'
import styled from 'styled-components/macro'

type RowSubtitleProps = typeof defaultProps & {
  children: React.ReactNode
}

const defaultProps = {}

const RowSubtitle = (props: RowSubtitleProps) => {
  const { children, ...rest } = props

  return <Title {...rest}>{children}</Title>
}

// TODO: Theme
const Title = styled.h4`
  color: black;
  font-size: 14rem;
  font-weight: 400;
  margin: 0;
`

RowSubtitle.defaultProps = defaultProps

export default RowSubtitle
