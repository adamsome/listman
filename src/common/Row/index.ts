export { default } from './Row'
export { default as RowContent } from './RowContent'
export { default as RowImage } from './RowImage'
export { default as RowTitle } from './RowTitle'
