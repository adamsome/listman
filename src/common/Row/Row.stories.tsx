import { action } from '@storybook/addon-actions'
import React from 'react'
import Starrate from 'react-minor-ui'
import Row from './Row'
import RowContent from './RowContent'
import RowImage from './RowImage'
import RowTitle from './RowTitle'
import sampleArtImage from './sample-art.png'

export default {
  component: Row,
  title: 'Row',
  excludeStories: ['actions', 'createRow'],
}

export const actions = {
  onClick: action('onClick'),
}

export const createRow = () => (
  <Row {...actions}>
    <RowImage src={sampleArtImage} height="120px" alt="Sample Art" />
    <RowContent>
      <RowTitle subtitle="Subtitle" actions={<Starrate />}>
        Title
      </RowTitle>
      Content
    </RowContent>
  </Row>
)

export const image = () => createRow()
export const lowWidth = () => (
  <div style={{ width: '400px' }}>{createRow()}</div>
)

// TODO: Story w/ long titles
