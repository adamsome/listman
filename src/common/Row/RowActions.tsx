import React from 'react'
import RowSection from './RowSection'

type Props = typeof defaultProps & {
  children: React.ReactNode
}

const defaultProps = {}

const RowActions = (props: Props) => {
  const { children, ...rest } = props

  return <RowSection {...rest}>{children}</RowSection>
}

RowActions.defaultProps = defaultProps

export default RowActions
