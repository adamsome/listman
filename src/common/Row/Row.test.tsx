import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import ThemeProvider from '../theming/ThemeProvider'
import { createRow } from './Row.stories'

const create = () => <ThemeProvider>{createRow()}</ThemeProvider>

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(create(), div)
  ReactDOM.unmountComponentAtNode(div)
})

it('renders correctly', () => {
  const tree = renderer.create(create()).toJSON()
  expect(tree).toMatchSnapshot()
})
