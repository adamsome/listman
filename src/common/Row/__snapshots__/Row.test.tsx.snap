// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.emotion-12 {
  color: hsl(0,0%,13%);
  background-color: hsl(0,0%,100%);
}

.emotion-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(252,12%,92%);
  border-bottom-width: 0;
  color: hsl(0,0%,13%);
  cursor: pointer;
  font-size: 1.000000rem;
  outline: none;
  overflow: hidden;
  text-align: start;
}

.emotion-11:first-of-type {
  border-top-left-radius: 6px;
  border-top-right-radius: 6px;
}

.emotion-11:last-of-type {
  border-bottom-left-radius: 6px;
  border-bottom-right-radius: 6px;
  border-bottom-width: 1px;
}

.emotion-11:focus {
  box-shadow: 0 0 0 2px hsla(206,71%,83%,0.5),0 1px 2px 0 rgba(0,0,0,0.2),0 2px 4px 0 rgba(0,0,0,0.2);
}

.emotion-0 {
  display: block;
  max-width: 100%;
  margin-left: 1rem;
  margin-right: 1rem;
}

.emotion-0:first-of-type {
  margin-left: 0;
}

.emotion-0:last-of-type {
  margin-right: 0;
}

.emotion-10 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: 1rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  overflow: hidden;
}

.emotion-10:first-of-type {
  margin-left: 1rem;
}

.emotion-9 {
  display: block;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-7:not(:last-of-type) {
  margin-bottom: 0.5rem;
}

.emotion-7:not(:first-of-type) {
  margin-top: 0.5rem;
}

.emotion-5 {
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  margin-right: 1rem;
  display: block;
  min-width: 0;
}

.emotion-2 {
  color: hsl(0,0%,13%);
  font-size: 1.142857rem;
  font-weight: 700;
  margin: 0;
}

.emotion-1 {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-4 {
  color: hsl(0,0%,13%);
  font-size: 1.142857rem;
  font-weight: 700;
  margin: 0;
  color: hsl(247,5%,67%);
}

.emotion-6 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.emotion-8 {
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

<div
  className="emotion-12"
>
  <div
    className="emotion-11"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <img
      alt="Sample Art"
      className="emotion-0"
      height="114px"
      src="sample-art.png"
    />
    <div
      className="emotion-10"
    >
      <div
        className="emotion-9"
      >
        <div
          className="emotion-7"
        >
          <div
            className="emotion-5"
          >
            <h4
              className="emotion-2"
            >
              <div
                className="emotion-1"
              >
                Long title of an type of artifact that is being rated
              </div>
            </h4>
            <h4
              className="emotion-4"
            >
              <div
                className="emotion-1"
              >
                Short subtitle
              </div>
            </h4>
          </div>
          <div
            className="emotion-6"
          >
            <div>
              <svg
                height="1.5rem"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                width="1.5rem"
              >
                <g
                  style={
                    Object {
                      "opacity": 0.25,
                      "transition": "opacity 100ms ease-in-out",
                    }
                  }
                  x={-0.2}
                  y={0.9}
                >
                  <svg
                    viewBox="0 0 512 512"
                  >
                    <circle
                      cx="256"
                      cy="256"
                      fill="currentColor"
                      r="40"
                      strokeWidth="0"
                      transform="inherit"
                      transform-origin="center"
                    />
                  </svg>
                </g>
                <g
                  style={
                    Object {
                      "opacity": 0,
                      "transform": "scale(0)",
                      "transformOrigin": "center",
                      "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                    }
                  }
                >
                  <svg
                    viewBox="0 0 51 48"
                  >
                    <path
                      d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                      fill="currentColor"
                      stroke="currentColor"
                    />
                  </svg>
                </g>
              </svg>
              <svg
                height="1.5rem"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                width="1.5rem"
              >
                <g
                  style={
                    Object {
                      "opacity": 0.25,
                      "transition": "opacity 100ms ease-in-out",
                    }
                  }
                  x={-0.2}
                  y={0.9}
                >
                  <svg
                    viewBox="0 0 512 512"
                  >
                    <circle
                      cx="256"
                      cy="256"
                      fill="currentColor"
                      r="40"
                      strokeWidth="0"
                      transform="inherit"
                      transform-origin="center"
                    />
                  </svg>
                </g>
                <g
                  style={
                    Object {
                      "opacity": 0,
                      "transform": "scale(0)",
                      "transformOrigin": "center",
                      "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                    }
                  }
                >
                  <svg
                    viewBox="0 0 51 48"
                  >
                    <path
                      d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                      fill="currentColor"
                      stroke="currentColor"
                    />
                  </svg>
                </g>
              </svg>
              <svg
                height="1.5rem"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                width="1.5rem"
              >
                <g
                  style={
                    Object {
                      "opacity": 0.25,
                      "transition": "opacity 100ms ease-in-out",
                    }
                  }
                  x={-0.2}
                  y={0.9}
                >
                  <svg
                    viewBox="0 0 512 512"
                  >
                    <circle
                      cx="256"
                      cy="256"
                      fill="currentColor"
                      r="40"
                      strokeWidth="0"
                      transform="inherit"
                      transform-origin="center"
                    />
                  </svg>
                </g>
                <g
                  style={
                    Object {
                      "opacity": 0,
                      "transform": "scale(0)",
                      "transformOrigin": "center",
                      "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                    }
                  }
                >
                  <svg
                    viewBox="0 0 51 48"
                  >
                    <path
                      d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                      fill="currentColor"
                      stroke="currentColor"
                    />
                  </svg>
                </g>
              </svg>
              <svg
                height="1.5rem"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                width="1.5rem"
              >
                <g
                  style={
                    Object {
                      "opacity": 0.25,
                      "transition": "opacity 100ms ease-in-out",
                    }
                  }
                  x={-0.2}
                  y={0.9}
                >
                  <svg
                    viewBox="0 0 512 512"
                  >
                    <circle
                      cx="256"
                      cy="256"
                      fill="currentColor"
                      r="40"
                      strokeWidth="0"
                      transform="inherit"
                      transform-origin="center"
                    />
                  </svg>
                </g>
                <g
                  style={
                    Object {
                      "opacity": 0,
                      "transform": "scale(0)",
                      "transformOrigin": "center",
                      "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                    }
                  }
                >
                  <svg
                    viewBox="0 0 51 48"
                  >
                    <path
                      d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                      fill="currentColor"
                      stroke="currentColor"
                    />
                  </svg>
                </g>
              </svg>
              <svg
                height="1.5rem"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                width="1.5rem"
              >
                <g
                  style={
                    Object {
                      "opacity": 0.25,
                      "transition": "opacity 100ms ease-in-out",
                    }
                  }
                  x={-0.2}
                  y={0.9}
                >
                  <svg
                    viewBox="0 0 512 512"
                  >
                    <circle
                      cx="256"
                      cy="256"
                      fill="currentColor"
                      r="40"
                      strokeWidth="0"
                      transform="inherit"
                      transform-origin="center"
                    />
                  </svg>
                </g>
                <g
                  style={
                    Object {
                      "opacity": 0,
                      "transform": "scale(0)",
                      "transformOrigin": "center",
                      "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                    }
                  }
                >
                  <svg
                    viewBox="0 0 51 48"
                  >
                    <path
                      d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                      fill="currentColor"
                      stroke="currentColor"
                    />
                  </svg>
                </g>
              </svg>
            </div>
          </div>
        </div>
        <div
          className="emotion-8"
        >
          No description.
        </div>
      </div>
    </div>
  </div>
</div>
`;
